orchagent/portsorch.cpp

```diff
@@ -463,6 +469,121 @@ bool PortsOrch::setPortFec(sai_object_id_t id, sai_port_fec_mode_t mode)
     return true;
 }
 
+bool PortsOrch::getPortPfc(sai_object_id_t portId, uint8_t *pfc_bitmask)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    Port p;
+
+    if (!getPort(portId, p))
+    {
+        SWSS_LOG_ERROR("Failed to get port object for port id 0x%lx", portId);
+        return false;
+    }
+
+    if (p.m_pfc_asym == SAI_PORT_PRIORITY_FLOW_CONTROL_MODE_COMBINED)
+    {
+        attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+    }
+    else
+    {
+        attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL_TX;
+    }
+
+    sai_status_t status = sai_port_api->get_port_attribute(portId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC value on port id 0x%lx (rc:%d)", portId, status);
+    }
+
+    *pfc_bitmask = attr.value.u8;
+
+    return true;
+}
+
+bool PortsOrch::setPortPfc(sai_object_id_t portId, uint8_t pfc_bitmask)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    Port p;
+
+    if (!getPort(portId, p))
+    {
+        SWSS_LOG_ERROR("Failed to get port object for port id 0x%lx", portId);
+        return false;
+    }
+
+    if (p.m_pfc_asym == SAI_PORT_PRIORITY_FLOW_CONTROL_MODE_COMBINED)
+    {
+        attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+    }
+    else
+    {
+        attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL_TX;
+    }
+
+    attr.value.u8 = pfc_bitmask;
+
+    sai_status_t status = sai_port_api->set_port_attribute(portId, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set PFC 0x%x to port id 0x%lx (rc:%d)", attr.value.u8, portId, status);
+        return false;
+    }
+
+    return true;
+}
+
+bool PortsOrch::setPortPfcAsym(Port &port, string pfc_asym)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    uint8_t pfc = 0;
+
+    if (!getPortPfc(port.m_port_id, &pfc))
+    {
+        return false;
+    }
+
+    port.m_pfc_asym = pfc_asym_map.at(pfc_asym);
```

Comment:
It will break if pfc_asym is not entered as “on/off”, please consider to handle this to make it more robust.